### PR TITLE
[NAE-1678] User impersonation

### PR DESCRIPTION
**Pull request by timbez**
# Description

- Implements a way for users to impersonate one another.

Implements [NAE-1678](https://netgrif.atlassian.net/browse/NAE-1678) User impersonation

## Dependencies

- none

### Third party dependencies

- none

### Blocking Pull requests

- will need conflict resolution with https://github.com/netgrif/application-engine/pull/54/

## How Has Been This Tested?

Tested manually (via Example app) and via ImpersonationServiceTest class (src/test/groovy/com/netgrif/application/engine/impersonation/ImpersonationServiceTest.groovy)

### Test Configuration

- no special parameters

| Name                | Tested on |
|---------------------| --------- |
| OS                  |     Win 10      |
| Runtime             |     Java 11      |
| Dependency Manager | Maven 3.8.4 |
| Framework version | Spring Boot 2.6.2 |

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] My changes have been checked, personally or remotely, with @...
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have resolved all conflicts with the target branch of the PR
- [x] I have updated and synced my code with the target branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes:
    - [x] Lint test
    - [x] Unit tests
    - [x] Integration tests
- [x] I have checked my contribution with code analysis tools:
    - [x] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [x] [Snyk](https://app.snyk.io/org/netgrif/project/ea82b3b8-658d-41f5-89a7-76cd600da01f)
- [ ] I have made corresponding changes to the documentation:
    - [ ] Developer documentation
    - [ ] User Guides
    - [ ] Migration Guides

### CODE REVIEW

File: src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        Page<Case> cases = findCases(request(impersonatorId, impersonatedId), PageRequest.of(0, MAX_CONFIGS_PER_USER));
+        return cases.getContent();
+    }
+
+    @Override
+    public List<Authority> getAuthorities(List<Case> configs, IUser impersonated) {
+        if (configs.isEmpty()) {
+            return new ArrayList<>();
+        }
+        List<String> authIds = (List) configs.get(0).getDataSet().get("impersonated_authorities").getValue();
```
Comment (minop) [inline src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java:96]: what about the authorities from the other config elements?

File: src/main/resources/petriNets/engine-processes/impersonation_config.xml [hunk 2]
```diff
@@ -0,0 +1,759 @@
...
+					<rows>1</rows>
+					<cols>4</cols>
+					<template>material</template>
+					<appearance>outline</appearance>
+				</layout>
+			</dataRef>
+		</dataGroup>
+		<event type="finish">
+			<id>finish</id>
+			<title>Publikovať</title>
```
Comment (minop) [inline src/main/resources/petriNets/engine-processes/impersonation_config.xml:471]: event lacks i18n

File: src/main/resources/petriNets/engine-processes/impersonation_users_select.xml [hunk 3]
```diff
@@ -0,0 +1,84 @@
+<document xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://petriflow.com/petriflow.schema.xsd">
+	<id>impersonation_users_select</id>
+	<version>1.0.0</version>
+	<initials>IPU</initials>
+	<title>Konfigurácia zastupiteľov - user select</title>
```
Comment (minop) [inline src/main/resources/petriNets/engine-processes/impersonation_users_select.xml:5]: i18n

**Review by minop — CHANGES_REQUESTED** [review]

File: src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java (hunk 1 above)
Comment (timbez) [reply src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java]: this is an oversight. This code branch is not actually used atm (frontend does not utilize this in this specific context) but I fixed it nonetheless

File: src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    @Override
+    public List<ProcessRole> getRoles(List<Case> configs, IUser impersonated) {
+        List<ProcessRole> impersonatedRoles = new ArrayList<>();
+        impersonatedRoles.add(processRoleService.defaultRole());
+        if (configs.isEmpty()) {
+            return impersonatedRoles;
+        }
+        List<String> roleIds = (List) configs.get(0).getDataSet().get("impersonated_roles").getValue();
```
[the remark this answers is not in the document]
Comment (timbez) [reply src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java]: same as above

File: src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    @Override
+    public boolean canImpersonate(LoggedUser impersonator, String configId) {
+        Case config = getConfig(configId);
+        return isValidAndContainsUser(config, impersonator.getId());
+    }
+
+    @Override
+    public boolean canImpersonateUser(LoggedUser impersonator, String userId) {
+        IUser impersonated = userService.findById(userId, true);
+        List<Case> impersonationConfig = searchConfigs(impersonator.getId(), impersonated.getStringId());
```
[the remark this answers is not in the document]
Comment (timbez) [reply src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java]: done

File: src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    private IProcessRoleService processRoleService;
+
+    @Override
+    public Page<IUser> getConfiguredImpersonationUsers(String query, LoggedUser impersonator, Pageable pageable) {
+        if (impersonator.isAdmin()) {
+            return userService.searchAllCoMembers(query, null, null, impersonator, true, pageable);
+
+        } else {
+            Page<Case> cases = searchConfigs(impersonator.getId(), pageable);
+            List<IUser> users = cases.getContent().stream()
```
[the remark this answers is not in the document]
Comment (timbez) [reply src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java:59]: pageable is passed into the method from the controller, so it is possible to continually request more results

File: src/main/java/com/netgrif/application/engine/workflow/service/DataService.java [hunk 7]
```diff
@@ -101,6 +101,7 @@ public GetDataEventOutcome getData(String taskId) {
     @Override
     public GetDataEventOutcome getData(Task task, Case useCase) {
         log.info("[" + useCase.getStringId() + "]: Getting data of task " + task.getTransitionId() + " [" + task.getStringId() + "]");
+        IUser user = userService.getLoggedOrSystem();
```
[the remark this answers is not in the document]
Comment (timbez) [reply src/main/java/com/netgrif/application/engine/workflow/service/DataService.java:104]: here the logged user object is only used to propagate its data into some event logs, their contructors handle the ids internally

File: src/main/java/com/netgrif/application/engine/workflow/service/DataService.java [hunk 8]
```diff
@@ -190,6 +191,7 @@ public SetDataEventOutcome setData(String taskId, ObjectNode values) {
     @Override
     public SetDataEventOutcome setData(Task task, ObjectNode values) {
         Case useCase = workflowService.findOne(task.getCaseId());
+        IUser user = userService.getLoggedOrSystem();
```
[the remark this answers is not in the document]
Comment (timbez) [reply src/main/java/com/netgrif/application/engine/workflow/service/DataService.java:194]: here the logged user object is only used to propagate its data into some event logs, their contructors handle the ids internally

File: src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java [hunk 9]
```diff
@@ -133,7 +133,7 @@ public AssignTaskEventOutcome assignTask(LoggedUser loggedUser, String taskId) t
         if (!taskOptional.isPresent())
             throw new IllegalArgumentException("Could not find task with id [" + taskId + "]");
 
-        IUser user = userService.resolveById(loggedUser.getId(), true);
+        IUser user = getUserFromLoggedUser(loggedUser);
```
[the remark this answers is not in the document]
Comment (timbez) [reply src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java:136]: the method is used down the line when the task execution/assign/cancel actually happens. The methods receive the full user object (actual user + optional nested impersonated user) so they have access to the full context; `getSelfOrImpersonated()` is then only called when necessary

File: src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java [hunk 10]
```diff
@@ -207,13 +207,13 @@ public FinishTaskEventOutcome finishTask(LoggedUser loggedUser, String taskId) t
         if (!taskOptional.isPresent())
             throw new IllegalArgumentException("Could not find task with id [" + taskId + "]");
         Task task = taskOptional.get();
-        IUser user = userService.resolveById(loggedUser.getId(), true);
+        IUser user = getUserFromLoggedUser(loggedUser);
```
[the remark this answers is not in the document]
Comment (machacjozef) [reply src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java:210]: https://github.com/netgrif/application-engine/pull/109/files#r982858760

File: src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java [hunk 11]
```diff
@@ -272,7 +272,7 @@ public CancelTaskEventOutcome cancelTask(LoggedUser loggedUser, String taskId) {
         if (!taskOptional.isPresent())
             throw new IllegalArgumentException("Could not find task with id [" + taskId + "]");
 
-        IUser user = userService.resolveById(loggedUser.getId(), true);
+        IUser user = getUserFromLoggedUser(loggedUser);
```
[the remark this answers is not in the document]
Comment (machacjozef) [reply src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java:275]: https://github.com/netgrif/application-engine/pull/109/files#r982858760

File: src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java [hunk 12]
```diff
@@ -346,7 +346,8 @@ private Task returnTokens(Task task, String useCaseId) {
     @Transactional
     public DelegateTaskEventOutcome delegateTask(LoggedUser loggedUser, String delegatedId, String taskId) throws TransitionNotExecutableException {
         IUser delegatedUser = userService.resolveById(delegatedId, true);
-        IUser delegateUser = userService.resolveById(loggedUser.getId(), true);
+        IUser delegateUser = getUserFromLoggedUser(loggedUser);
```
[the remark this answers is not in the document]
Comment (machacjozef) [reply src/main/java/com/netgrif/application/engine/workflow/service/TaskService.java:349]: https://github.com/netgrif/application-engine/pull/109/files#r982858760

File: src/main/groovy/com/netgrif/application/engine/startup/ImpersonationRunner.groovy [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+@Slf4j
+@Component
+class ImpersonationRunner extends AbstractOrderedCommandLineRunner {
+
+    @Autowired
+    private IPetriNetService petriNetService
```
Comment (machacjozef) [inline src/main/groovy/com/netgrif/application/engine/startup/ImpersonationRunner.groovy:15]: protected 

(tutok a vsade totok)

File: src/main/java/com/netgrif/application/engine/configuration/security/ImpersonationRequestFilter.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        try {
+            log.debug("Filtering request " + servletRequest.getRequestURI() + ", " + loggedUser.getUsername());
+            impersonatorRepository.findByImpersonatedId(loggedUser.getId()).ifPresent(impersonatorRepository::delete);
+
+        } catch (Exception e) {
+            log.error("Failed to resolve impersonators for " + loggedUser.getUsername() + ", " + e.getMessage(), e);
+        }
+    }
+
+    private boolean isValid(Impersonator impersonator) {
```
Comment (machacjozef) [inline src/main/java/com/netgrif/application/engine/configuration/security/ImpersonationRequestFilter.java:71]: protected

**Review by machacjozef — CHANGES_REQUESTED** [review]

File: src/main/resources/petriNets/engine-processes/impersonation_config.xml [hunk 15]
```diff
@@ -0,0 +1,759 @@
...
+	</caseEvents>
+	<role>
+		<id>admin_imp</id>
+		<title>Admin</title>
+	</role>
+	<role>
+		<id>user_imp</id>
+		<title>Používateľ</title>
+	</role>
+	<data type="text" immediate="true">
```
[the remark this answers is not in the document]
Comment (machacjozef) [reply src/main/resources/petriNets/engine-processes/impersonation_config.xml:126]: done

File: src/main/resources/petriNets/engine-processes/impersonation_config.xml [hunk 16]
```diff
@@ -0,0 +1,759 @@
...
+	</data>
+	<data type="taskRef">
+		<id>user_select_taskRef</id>
+		<title/>
+	</data>
+	<data type="text">
+		<id>user_select_caseId</id>
+		<title/>
+	</data>
+	<transition>
```
[the remark this answers is not in the document]
Comment (machacjozef) [reply src/main/resources/petriNets/engine-processes/impersonation_config.xml:350]: done

File: src/main/resources/petriNets/engine-processes/impersonation_config.xml (hunk 2 above)
Comment (machacjozef) [reply src/main/resources/petriNets/engine-processes/impersonation_config.xml]: done

File: src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (minop) [reply src/main/java/com/netgrif/application/engine/impersonation/service/ImpersonationAuthorizationService.java:61]: bot how does the caller know, that there are more pages left?

File: src/main/resources/petriNets/engine-processes/impersonation_users_select.xml [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+		<id>parentId</id>
+		<title/>
+	</data>
+	<data type="enumeration_map">
+		<id>select_impersonator</id>
+		<title name="select_impersonator">Select impersonator</title>
+		<component>
+			<name>autocomplete_dynamic</name>
+		</component>
+		<placeholder>Vyhľadajte podľa mena alebo emailovej adresy</placeholder>
```
Comment (minop) [inline src/main/resources/petriNets/engine-processes/impersonation_users_select.xml:29]: i18n

File: src/main/resources/petriNets/engine-processes/impersonation_config.xml [hunk 18]
```diff
@@ -0,0 +1,797 @@
...
+		<i18nString name="impersonators">Konfigurierte Vertreter</i18nString>
+		<i18nString name="valid_from">Gültig ab</i18nString>
+		<i18nString name="valid_to">Gültig bis</i18nString>
+		<i18nString name="is_active">Aktive Konfiguration</i18nString>
+		<i18nString name="impersonated">Benutzer zu vertreten</i18nString>
+		<i18nString name="remove_impersonator">Ausgewählte löschen</i18nString>
+		<i18nString name="add_selected">Ausgewählte hinzufügen</i18nString>
+		<i18nString name="impersonated_authorities">Vertretene Berechtigungen</i18nString>
+		<i18nString name="impersonated_roles">Vertretene Rollen</i18nString>
+		<i18nString name="configure_impersonators">Vertreter wählen</i18nString>
```
Comment (minop) [inline src/main/resources/petriNets/engine-processes/impersonation_config.xml:54]: Vertreter auswählen

**Review by minop — CHANGES_REQUESTED** [review]

**Review by minop — APPROVED** [review]

**Review by mladoniczky — APPROVED** [review]

**Pull request merged**